panther_lights/test/test_controller_node.cpp

```diff
@@ -0,0 +1,325 @@
+}
+
+TEST_F(TestControllerNode, LoadAnimationInvalidPriority)
+{
+  YAML::Node led_animation_desc;
+  led_animation_desc["id"] = 11;
+  led_animation_desc["priority"] = 0;
+
+  EXPECT_TRUE(panther_utils::test_utils::IsMessageThrown<std::runtime_error>(
+    [&]() { controller_node_->LoadAnimation(led_animation_desc); },
+    "Invalid LED animation priority"));
+
+  led_animation_desc["priority"] = 4;
+
+  EXPECT_TRUE(panther_utils::test_utils::IsMessageThrown<std::runtime_error>(
+    [&]() { controller_node_->LoadAnimation(led_animation_desc); },
+    "Invalid LED animation priority"));
+}
+
+TEST_F(TestControllerNode, LoadAnimationThrowRepeatingID)
+{
+  YAML::Node led_animation_desc;
+  led_animation_desc["id"] = 11;
+  led_animation_desc["animations"] = std::vector<YAML::Node>();
+
+  ASSERT_NO_THROW(controller_node_->LoadAnimation(led_animation_desc));
+
+  EXPECT_TRUE(panther_utils::test_utils::IsMessageThrown<std::runtime_error>(
+    [&]() { controller_node_->LoadAnimation(led_animation_desc); },
+    "Animation with given ID already exists"));
+}
+
+TEST_F(TestControllerNode, AddAnimationToQueueThrowBadAnimaitonID)
+{
+  EXPECT_TRUE(panther_utils::test_utils::IsMessageThrown<std::runtime_error>(
+    [&]() { controller_node_->AddAnimationToQueue(99, false); }, "No animation with ID:"));
+}
+
+TEST_F(TestControllerNode, AddAnimationToQueue)
+{
+  auto queue = controller_node_->GetQueue();
+  EXPECT_TRUE(queue->Empty());
+  EXPECT_NO_THROW(controller_node_->AddAnimationToQueue(0, false));
+  EXPECT_FALSE(queue->Empty());
+}
+
+TEST_F(TestControllerNode, CallSelLEDAnimaitonService)
+{
+  this->CallSetLEDAnimationSrv(0, false);
+
+  // spin to invoke timer
+  auto anim = controller_node_->GetCurrentAnimation();
+  while (!anim) {
+    rclcpp::spin_some(controller_node_->get_node_base_interface());
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    anim = controller_node_->GetCurrentAnimation();
+  }
+
+  EXPECT_STREQ("ANIMATION_0", anim->GetName().c_str());
+
+  // add another animation to queue
+  auto queue = controller_node_->GetQueue();
+  EXPECT_TRUE(queue->Empty());
+
+  this->CallSetLEDAnimationSrv(0, false);
+
+  EXPECT_FALSE(queue->Empty());
+}
+
+TEST_F(TestControllerNode, CallSelLEDAnimaitonServicePriorityInterrupt)
+{
+  this->CallSetLEDAnimationSrv(0, false);
+
+  // spin to invoke timer
+  auto anim = controller_node_->GetCurrentAnimation();
+  while (!anim) {
+    rclcpp::spin_some(controller_node_->get_node_base_interface());
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    anim = controller_node_->GetCurrentAnimation();
+  }
+
+  // add animation with higher priority spin to give time for controller to overwrite current
+  // animation then check if animation has changed
+  this->CallSetLEDAnimationSrv(1, false);
+  for (int i = 0; i < 10; i++) {
+    rclcpp::spin_some(controller_node_->get_node_base_interface());
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  }
```

Comment:
How about creating a utility function for spinning the node until the message received/shared_ptr is not nullptr, or timeout occurred?

We had sth similar in one of our previous projects and it did well.

Reply:
This cases are quite unique as we need to spin node to check if one of class objects state has changed. This would require a method that spins the node and checks the condition passed as parameter with timeout I guess. Not sure how useful this utility will be that's why I left it this way.

Reply:
I do not quite understand the problem and moving main function to the bottom of the file shouldn't influence the test behaviour.